compel/src/lib/infect.c

```diff
@@ -856,15 +854,47 @@ void compel_relocs_apply(void *mem, void *vbase, struct parasite_blob_desc *pbd)
 #endif
 }
 
+long remote_mprotect(struct parasite_ctl *ctl, void *addr, size_t len, int prot)
+{
+	long ret;
+	int err;
+	bool compat_task = !user_regs_native(&ctl->orig.regs);
+
+	err = compel_syscall(ctl, __NR(mprotect, compat_task), &ret,
+			(unsigned long)addr, len, prot, 0, 0, 0);
+	if (err < 0) {
+		pr_err("compel_syscall for mprotect failed\n");
+		return -1;
+	}
+	return ret;
+}
+
 static int compel_map_exchange(struct parasite_ctl *ctl, unsigned long size)
 {
-	int ret;
+	int ret, remote_prot;
 
-	ret = parasite_memfd_exchange(ctl, size);
+	if (ctl->pblob.hdr.data_off)
+		remote_prot = PROT_READ | PROT_EXEC;
+	else
```

Comment:
Will it ever take the `else` arm?
`data_off` will be set to offset of `.data` which is always present, isn't it?

Reply:
Yes, for example on arm and aarch64 the `.data` section is inside another section (`.crblob`), so the `handle-elf.c` doesn't find it.  On mips it looks like `.data` is inside `.text`? ppc64, s390 and x86 looks like I'd expect them to. So I could align it on ppc64, s390 and x86 while on the other architectures it would work as before. What do you think?

Reply:
I think it could work, just ppc64 would need `.data` aligned at 64Kb as this is the default page size on ppc64.
Yet, arm and arm64 may also have kernels with W^X enabled, so it would be nice to convert them as well :)

Reply:
Done for ppc64 and s390. Unfortunately I don't understand this code well enough to implement it for arm (especially without an access to it).